### PR TITLE
bpo-41930: Add support for SQLite serialise/deserialise API

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue41930

### CODE REVIEW

**Comment by erlend-aasland** [comment]
FYI, the ASAN failure in `test_multiprocessing_fork` looks unrelated.

**Comment by erlend-aasland** [comment]
FYI, I'll resolve the What's New entry after the buildbots are done.

**Comment by erlend-aasland** [comment]
Here's the buildbot summary for 5aa2ec54f232e1053f9ec2f94cad78d1c9576eb5:
_71 successful, 1 skipped, and 2 failing checks_

The two failures were:
1. Tests/Address sanitizer: `test_get (test.test_multiprocessing_fork.WithProcessesTestQueue) ... Timeout (0:20:00)`
2. buildbot/AMD64 Arch Linux Asan Debug PR: `RuntimeError: subprocess not supported for isolated subinterpreters` (while compiling the `_struct` extension module)

**Comment by erlend-aasland** [comment]
Windows failure is unrelated (asyncio failed at `test_sendfile_close_peer_in_the_middle_of_receiving`)

**Comment by erlend-aasland** [comment]
Note to self: on hold until GH-29507 has landed; use `configure` to autodetect the presence of `sqlite3_serialize` and produce a define we can wrap this in.

**Comment by erlend-aasland** [comment]
@serhiy-storchaka Are you fine with adding these API's?

File: Doc/library/sqlite3.rst
Comment (JelleZijlstra) [inline Doc/library/sqlite3.rst]:
```suggestion
      This method serializes a database into a :class:`bytes` object.  For an
```
Your wording sounds like it returns a list of bytes objects.

File: Doc/library/sqlite3.rst
Comment (JelleZijlstra) [inline Doc/library/sqlite3.rst]:
```suggestion
      disk if that database were backed up to disk.
```

File: Modules/_sqlite/connection.c
Comment (JelleZijlstra) [inline Modules/_sqlite/connection.c]:
```suggestion
were backed up to disk.
```

File: Modules/_sqlite/connection.c
Comment (JelleZijlstra) [inline Modules/_sqlite/connection.c]:
```suggestion
    if (data->len >= (2 << 62)) {
```
Or some other variation that's not just a magic number.

File: pyconfig.h.in
Comment (JelleZijlstra) [inline pyconfig.h.in]:
```suggestion
/* Define if SQLite was compiled with the serialize API */
```

**Review by JelleZijlstra — COMMENTED** [review]
Took a first look and noticed some small things.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Thanks!

File: pyconfig.h.in
Comment (erlend-aasland) [reply pyconfig.h.in]: FYI, this is generated by `AC_DEFINE` in `configure.ac`.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Seems like this broke the Windows builds; I think just writing it out literally may be the best option, to avoid compiler tricks. There also seems to be some precedent for writing these out explicitly:

```
$ grep -r 9223372036854775807 * | wc -l
      27
$ grep -r 9223372036854775808 * | wc -l
      19
$ grep -r "2 << 62" * | wc -l          
       1
```

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I added a comment instead; it seems that the MSVC compiler does some tricks here (I tried to find out via godbolt.org, but to no avail; perhaps explicitly marking the shifted literal as `long long` would have worked...) The code base also seem to prefer the explicit to the implicit here:

```
$ grep -r "922337203685477580[78]" * | wc -l
      47
$ grep -r "[12] << 6[23]" * | wc -l          
      13
```

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: On MSVC, `(1 << 63) - 1` triggers [MSVC compiler warning C4293](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4293?view=msvc-170). Using `(1LL << 63) - 1` seems to fix it, but the MS docs promotes casting to `unsigned __int64`.

I prefer just writing it out instead; also, there is already a comment explaining why we do this :)

```
     * Make sure we don't overflow sqlite3_deserialize(); it accepts a signed
     * 64-bit int as its data size argument, but sqlite3_malloc64 accepts an
     * unsigned 64-bit int as its size argument.
```

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: (Actually, I can reduce that comment; only the first part is important.)

File: Modules/_sqlite/connection.c
Comment (JelleZijlstra) [inline Modules/_sqlite/connection.c:1853]: reference: https://www.sqlite.org/c3ref/serialize.html

It sounds like the `SQLITE_SERIALIZE_NOCOPY` could provide a useful optimization, since we end up immediately copying the memory into a new bytes object. However, it would be more complex to use; most likely we'd have to first try with `SQLITE_SERIALIZE_NOCOPY` on, then if that fails, try again with it off. I don't know if that's worth the complexity, and in any case we're free to change it later if a user complains about the speed of repeatedly serializing a database.

File: Modules/_sqlite/connection.c
Comment (JelleZijlstra) [inline Modules/_sqlite/connection.c]: What does "non-standard" mean here? I see it on a lot of other docstrings in this file, but SQLite's own documentation doesn't say anything similar.

File: Modules/_sqlite/connection.c
Comment (JelleZijlstra) [inline Modules/_sqlite/connection.c:1928]: I had to convince myself the memory handling here was right. Spoiler alert: it is. Reference https://www.sqlite.org/c3ref/deserialize.html.

We set the FREEONCLOSE flag, which means that sqlite will treat the data buffer as its own, and free it when the database is closed. It will also free it when the function fails, so it's correct that the error case on line 1921  below doesn't free anything.

(No changes requested)

File: Modules/_sqlite/connection.c
Comment (JelleZijlstra) [inline Modules/_sqlite/connection.c]: Should we return None instead since the function never returns False?

File: Doc/library/sqlite3.rst
Comment (JelleZijlstra) [inline Doc/library/sqlite3.rst]:
```suggestion
         deserialize API.
```

File: Doc/library/sqlite3.rst
Comment (JelleZijlstra) [inline Doc/library/sqlite3.rst]: That's true at the C level, but for the Python-level documentation it will be more useful to say what exception it will throw.

Generally I noticed that the documentation is partially copied from SQLite's own docs. That should be fine because SQLite is in the public domain and obviously it's a good source for documentation about SQLite itself.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks. I'm on board with merging this, since it's generally useful for sqlite3 to expose more of the underlying SQLite API, the feature is simple and shouldn't need much maintenance in the future, and there haven't been any objections on the issue.

I do have a few small comments. Once those are addressed, I'll ask Guido for a second look and then merge the PR.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: It's a historic artefact; Gerhard marked all non-PEP-249 methods with "non-standard". Let's just remove those in another PR; I know it is confusing to some (most?) people.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:1928]: Yeah, it took me a little while to get that right; I misread the docs first, so the first iteration of this PR had it all wrong 😆

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Yes, we should.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:1853]: Let's try!

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: > Generally I noticed that the documentation is partially copied from SQLite's own docs. That should be fine because SQLite is in the public domain and obviously it's a good source for documentation about SQLite itself.

Yes, your observation is correct.

> That's true at the C level, but for the Python-level documentation it will be more useful to say what exception it will throw.

Yes, I'll reword it.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:1853]: See 7edccfed418eca5cdf88517cf9f3e91c2a051bc8.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: See GH-31612

**Review by JelleZijlstra — APPROVED** [review]
Looks good! @gvanrossum I'm planning to merge this PR (adding two new functions to sqlite3)

**Comment by gvanrossum** [comment]
> Looks good! @gvanrossum I'm planning to merge this PR (adding two new functions to sqlite3)

SGTM

File: Doc/library/sqlite3.rst
Comment (kumaraditya303) [inline Doc/library/sqlite3.rst:753]:
```suggestion
      Serializes a database into a :class:`bytes` object.  For an
```

File: PCbuild/_sqlite3.vcxproj
Comment (kumaraditya303) [inline PCbuild/_sqlite3.vcxproj]:
```suggestion
      <PreprocessorDefinitions>PY_SQLITE_HAVE_SERIALIZE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```

Comment (erlend-aasland) [reply PCbuild/_sqlite3.vcxproj]: Thanks!

**Comment by erlend-aasland** [comment]
FYI: synced with `main` and regenerated clinic bco. GH-31612.

**Comment by erlend-aasland** [comment]
@JelleZijlstra, will you be able to merge this before todays alpha?

**Comment by JelleZijlstra** [comment]
Oops, sorry for dropping this. There are merge conflicts on the configure scripts, could you resolve those?

**Review by JelleZijlstra — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, Jelle!

**Pull request merged**
